src/ziplist.c

```diff
@@ -1265,6 +1265,42 @@ unsigned char *ziplistDeleteRange(unsigned char *zl, int index, unsigned int num
     return (p == NULL) ? zl : __ziplistDelete(zl,p,num);
 }
 
+/* Replaces the entry at p. This is equivalent to a delete and an insert,
+ * but avoids some overhead when replacing av value of the same size. */
+unsigned char *ziplistReplace(unsigned char *zl, unsigned char *p, unsigned char *s, unsigned int slen) {
+
+    /* get metadata of the current entry */
+    zlentry entry;
+    zipEntry(p, &entry);
+
+    /* compute length of entry to store, excluding prevlen */
+    unsigned int reqlen;
+    unsigned char encoding = 0;
+    long long value = 123456789; /* initialized to avoid warning. */
+    if (zipTryEncoding(s,slen,&value,&encoding)) {
+        reqlen = zipIntSize(encoding); /* encoding is set */
+    } else {
+        reqlen = slen; /* encoding == 0 */
+    }
+    reqlen += zipStoreEntryEncoding(NULL,encoding,slen);
+
+    if (reqlen == entry.lensize + entry.len) {
```

Comment:
Good idea. I will add unit tests for replace.

Yes, they are tested in unit/type/hash.tcl. There are tests for both same size (HINCRY 100 to 102, both 1 byte integer encoded) and different size (HINCRBYFLOAT from 100 (1 byte integer) to 102.5 (5 byte string)).

Btw, is there a way to generate a coverage report?

Reply:
i saw `make gcov`, never tried it

Reply:
Nice! I found out `make lcov` is the command to use. It compiles with gcov, runs the tests and produces HTML reports and very nice tables. I had to install `apt install lcov`. The `ziplistReplace` function is covered 388768 times, of which 338341 benefit from the optimization and only 18818 go into the fallback else clause (delete + insert), so if real data looks like the unit test data, this is pretty good.

Reply:
I haven't found out how to run the unit tests inside ziplist.c yet though...

Reply:
I don't think there is a build argument for it, but you need to define REDIS_TEST and this it can be executed with redis-server test ziplist for this: https://github.com/redis/redis/blob/unstable/src/server.c#L5906